src/esi/Expression.cc

```diff
@@ -63,15 +63,15 @@ typedef enum {
 } literalhint;
 
 struct _stackmember {
-    evaluate *eval;
+    evaluate *eval = nullptr;
     union {
-        char *string;
+        char *string = nullptr;
         double floating;
         int integral;
     } value;
```

Comment:
This implies that sizeof(double) <= sizeof(char*) and also opens up a possibility that a future bigger union member will be half-initialized by mistake. From a defensive coding point of view, we should do this instead AFAICT:

```suggestion
        char *string;
        double floating;
        int integral;
    } value = {};
```

Reply:
We should be able to verify the situation with static_assert as well if there is any doubt.

Reply:
I found the following [cppreference](https://en.cppreference.com/w/cpp/language/aggregate_initialization) statement that invalidates my specific suggestion: "When a union is initialized by aggregate initialization, only its first non-static data member is initialized". That is not what GCC did in my test[^1], but that does not matter -- we should go by the standard, of course. My suggested ` = {}` change is worse than the current PR code because the latter is _explicitly_ flawed.

[^1]: I did that test before making my bad suggestion because I could not find the right standard promise at the time. GCC is probably buggy in this area as discussed on [SA](https://stackoverflow.com/questions/68881411/aggregate-initialization-of-a-union-in-c-with). The adjusted test (currently set to use clang) is [available](https://cpp.godbolt.org/z/ocqaz7s6T) if you want to play with it.

AFAICT, fixing this needs an explicit union constructor that memsets the whole union[^2]. There is no other way to do it in C++.

[^2]: That was actually my first idea, but then I saw other folks using aggregate initialization for unions and, after validating with an evidently buggy GCC, succumbed to the temptation :-(.

> Amos: We should be able to verify the situation with static_assert as well if there is any doubt.

I do not see how it is possible to reliably verify runtime zero-initialization with a static assert but would be glad to be proven wrong. With an explicit memset() solution, we would not have to double check.

Reply:
> 
> > Amos: We should be able to verify the situation with static_assert as well if there is any doubt.
> 
> I do not see how it is possible to reliably verify runtime zero-initialization with a static assert 

I was thinking the below, but it does require that GCC initializer-list behaviour:
```
static_assert(Foo().value.string == nullptr);
static_assert(Foo().value.floating == 0.0);
static_assert(Foo().value.integral == 0);
```
clang complains about "active member" being different from the "accessed member" on the latter two.

Reply:
The static assert you were thinking about is not reliable enough to be worth using IMO: That assert may succeed[^1] in test environments where we want it to fail.

[^1]: It succeed when the memory behind the temporary Foo() object is zeroed for other reasons, which is often the case. If you have seen my [test](https://cpp.godbolt.org/), you know what kind of hoops we need to jump through to test this initialization reliably, but that testing approach is probably too cumbersome or impossible for a static assert. Fortunately, there are no reasons to double check a basic memset() result!

> clang complains about "active member" being different from the "accessed member" on the latter two.

That is a side effect of broken initialization. That specific complaint will go away when the initialization is fixed (or if the assert expression is refactored), but that complaint is not the reason why we should not use that static assert.

Reply:
@kinkie, please let me know if you want me to take a stab at fixing this using memset(). I do not know whether my verbal description was good enough, and the changes require touching more lines than a GitHub suggestion for this change request can express...

Reply:
I've taken a stab at it